src/rulesets/Base/mapreduce.jl

```diff
@@ -48,3 +48,82 @@ function rrule(
     end
     return y, sum_abs2_pullback
 end
+
+#####
+##### `prod`
+#####
+
+function rrule(::typeof(prod), x::AbstractArray{T}; dims=:) where {T<:CommutativeMulNumber}
```

Comment:
Apologies for diving in with my usual request, but is there a sensible way that we could restrict the type here, since the tests currently only look at `Array`s? e.g. I imagine that a at least one of a `Fill`, `Diagonal`, `StaticArray` etc will do something weird here. Would `StridedArray` suffice for your use case?

Reply:
There is a test with PermutedDimsArray, which isn't a StridedArray, and I think it ought to work fine with StaticArrays, although I have not tested that in any depth. Diagonal seems to work although I struggle to imagine why calling `prod` on one would be a good idea, but weird things happen:
```
julia> unthunk(rrule(prod, Diagonal(SA[1,2,3,4]))[2](1.0)[2])
4×4 Diagonal{Float64, MVector{4, Float64}}:
 0.0   ⋅    ⋅    ⋅ 
  ⋅   0.0   ⋅    ⋅ 
  ⋅    ⋅   0.0   ⋅ 
  ⋅    ⋅    ⋅   0.0
  
julia> unthunk(rrule(prod, Fill(2,3))[2](1.0)[2])
3-element Vector{Float64}:
 4.0
 4.0
 4.0  
```
Fill makes a Vector gradient. Somehow `rrule(sum, Fill(2,3))` makes a Fill, because it simply broadcasts rather than calling `similar`. Is this something the package aims to guarantee? I don't see a test for it.

Reply:
> There is a test with PermutedDimsArray, which isn't a StridedArray

Apologies, it's late here. They are quite clearly in the tests...

> Fill makes a Vector gradient

We would definitely want the output of `rrule` w.r.t. a `Fill` argument to be either another `Fill` or an appropriate `Composite`. This is the kind of thing that e.g. `Zygote` can probably get right without a rule in `ChainRules`, so I think the ideal solution here is just not to implement a rule that covers `Fill`.

Also, should the result with `Diagonal` have zeros on the diagonal?

Reply:
> Also, should the result with Diagonal have zeros on the diagonal?

Something is broken on the last commit, but not this, this one you can do in your head: It's a product of mostly zeros, so the gradient with respect to nonzero entries still vanishes. 

>>  Fill makes a Vector gradient
> 
> We would definitely want 

This could be arranged at the cost of more complexity... although possibly Fill ought to define `similar` more like that of Diagonal if it wishes to be preserved under such things? 

Although clearly not all gradients are going to perserve this structure:
```
julia> gradient(sum∘cumsum, Fill(1,3))[1]
3-element Vector{Int64}:
 3
 2
 1
 ```
 And here's how much Zygote can figure out right now:
 ```
 julia> function myprod(xs)
       out = 1.0
       for x in xs
         out *= x
       end
       out
       end
myprod (generic function with 1 method)

julia> gradient(myprod, Fill(1,3))[1]
3-element Vector{Float64}:
 1.0
 1.0
 1.0
```

Reply:
> Something is broken on the last commit, but not this, this one you can do in your head: It's a product of mostly zeros, so the gradient with respect to nonzero entries still vanishes.

Yes, good point.

> julia> gradient(sum∘cumsum, Fill(1,3))[1]

This should produce a `Composite`, with the `value` with an appropriate `value` field.

>  julia> gradient(x -> sum(cumsum(x, dims=1)), Diagonal([1,2,3]))[1]  # what should this produce?

This should also produce either a `Diagonal` or an appropriate `Composite`.

But with all of these types, I'm not saying that your PR needs to cover them. I'm purely suggesting it addresses the minimal set of types that you're confident are done correctly, and assumes that AD can do a reasonable job of deriving the others.

> julia> gradient(myprod, Fill(1,3))[1]

The answer to this is the result of a bug in `Zygote` that I should fix -- it looks like an example of what I'm commenting on here, where `getindex` has been implemented for too broad a set of types. `Zygote` really _should_ be able to derive the rule for this properly. i.e. `getindex` only ever returns the `value` field of a `Fill`, so you shouldn't even need a rule for `getindex` for `Fill`.

Reply:
So we can get this merged, shall we change this to `StridedArray` and then we can make a follow up later?

Reply:
I think everything in https://github.com/JuliaDiff/ChainRules.jl/pull/337#issuecomment-832080195 applies here. 

When last I checked there were weird test errors maybe with zero-dim arrays.